internal/wclayer/baselayerreader.go

```diff
@@ -0,0 +1,218 @@
+package wclayer
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/internal/longpath"
+	"github.com/Microsoft/hcsshim/internal/oc"
+	"go.opencensus.io/trace"
+)
+
+type baseLayerReader struct {
+	ctx          context.Context
+	s            *trace.Span
+	root         string
+	result       chan *fileEntry
+	proceed      chan bool
+	currentFile  *os.File
+	backupReader *winio.BackupFileReader
+}
+
+func newBaseLayerReader(ctx context.Context, root string, s *trace.Span) (r *baseLayerReader) {
+	r = &baseLayerReader{
+		ctx:     ctx,
```

Comment:
is this needed? I dont see it called anywhere, and i would prefer avoiding storing contexts in structs
(I know its done everywhere else, but we may as well change that here)

Reply:
I _suspect_ this was intended to be used in a `ctx.cancelled()` check; I recall adding (or copying in from somewhere) a similar thing to backuptar since containerd was cancelling contexts passed into long-running processes when layer operations were aborted.

Perhaps the lambda passed to `filepath.Walk` should raise an error if it finds `r.ctx.cancelled()`? Otherwise whatever loop is calling `Next` (I forget where) would need to be checking this.

Reply:
It also might make more sense to pass `ctx` to `r.walk` rather than storing it in the structure, as it's not used by code _outside_ that goroutine, and it seems more idiomatic.

I also find myself wondering if there was going to be an OC subspan here or something to separate the goroutine from the parent span. Although in this case the goroutine should be fully encompassed by `r:= newBaseLayerReader`=>`r.Close`. Perhaps something like `oc.event` for each file would be useful (but verbose!) for debugging.

Reply:
I think im of the same mindset: passing `ctx` to `walk` feels cleaner.
I also am confused by the span, since its closed both in `Close`, and by the caller.
I might make sense to pass that to `walk` as well so that `walk` can create its own space that lasts for the entire go routine -- but that probably should be saved for updating all of `"wclayer"` wholesale...

Reply:
To summarize a few action items:

  * The ```ctx``` is currently not used anywhere meaningful, and could be removed
  * The  ```legacyLayerReader{}``` doesn't seem to use it either
  * The current ```baseLayerReader{}``` is used in ```NewLayerReader()``` which doesn't seem to make use of the ctx for anything other than creating a span.
  * The layer reader is used in ```ExportLayerToTar()``` which calls ```LayerReader.Close()``` if the context is canceled. 

If we are to make use of the context in the layer readers (both the legacy one and the base layer reader), we need to re-work how they are used. I think this should be part of a different effort.

I can remove the ctx from the base layer exporter for now. How we decide to refactor later, we can bring up in a different issue/PR. What do you folks think?

Reply:
I agree, the refactoring can happen later, and shouldn't block this